source/Ship.cpp

```diff
@@ -763,6 +766,171 @@ void Ship::FinishLoading(bool isNewInstance)
+	if(CanBeCarried())
+		return;
+	const vector<Hardpoint> &hardpoints = Weapons();
+	// When loading or reloading bay indexes are not persisted.  Scrub all bays
+	// associated with weapons so that they can be reconfigured.
+	ScrubWeaponBays();
+	for(unsigned i = 0; i < hardpoints.size(); i++)
+	{
+		const Weapon *weapon = hardpoints[i].GetOutfit();
+		if(weapon && weapon->IsBay())
+			AddBay(hardpoints[i], i);
+		else
+			RemoveBay(hardpoints[i], i);
+	}
+}
+
+
+
+// Upon loading bays may have duplicates.  This wipes out all weapon bays
+// associated with weapon hardpoints before configuring them.
+void Ship::ScrubWeaponBays()
+{
+	const vector<Hardpoint> &hardpoints = Weapons();
+	Point p;
+	// It is possible to start with no bays so clear vector.
+	if(!HasBays())
+		bays.clear();
+	for(unsigned i = 0; i < hardpoints.size(); i++)
+	{
+		p = hardpoints[i].GetPoint();
+		for(auto it = bays.begin(); it != bays.end(); )
+		{
+			Bay &bay = *it;
+			if(bay.point.X() != p.X() || bay.point.Y() != p.Y())
+			{
+				++it;
+				continue;
+			}
+			it = bays.erase(it);
+		}
+	}
+	// It is possible to end with no bays so clear vector.
+	if(!HasBays())
+		bays.clear();
+}
+
+
+
+void Ship::SwapBay(int first, int second)
+{
+	Hardpoint hardpoint = Weapons()[first];
+	Hardpoint secondHardpoint = Weapons()[second];
+	Hardpoint newHardpoint = hardpoint;
+	const Weapon *weapon = hardpoint.GetOutfit();
+	const Weapon *secondWeapon = secondHardpoint.GetOutfit();
+	// No swap needs to happen if both are drone bays.
+	if(weapon && secondWeapon && weapon->IsBay() && secondWeapon->IsBay())
+		return;
+	// No swap needs to happen if there's no drone bay involved.
+	if(!(weapon && weapon->IsBay()) && !(secondWeapon && weapon->IsBay()))
+		return;
+	Point newPoint;
+	long unsigned int newIndex;
+	long unsigned int oldIndex;
+	if(weapon && weapon->IsBay())
+	{
+		newIndex = first;
+		oldIndex = second;
+	}
+	else
+	{
+		newHardpoint = secondHardpoint;
+		newIndex = second;
+		oldIndex = first;
+	}
+	for(unsigned int i = 0; i < bays.size(); i++)
+	{
+		if(bays[i].bayIndex == oldIndex)
+		{
+			UpdateBay(bays[i], newHardpoint, newIndex);
+			break;
+		}
+	}
+}
+
+
+
+// Add a drone or fighter bay to the ship (comes from bay outfit).
+// TODO: fix orientation
+void Ship::AddBay(Hardpoint hardpoint, long unsigned int bayIndex)
+{
+	const Weapon *weapon = hardpoint.GetOutfit();
+	// Only add a drone bay if the outfit is for drone bays.
+	if(!weapon || !weapon->IsBay())
+		return;
+	string category = "Drone";
+	if(hardpoint.IsTurret())
+		category = "Fighter";
+	// Only update drone bays for drones or fighter bays for fighters
+	if((category == "Drone" && !weapon->IsDroneBay()) || (category == "Fighter" && !weapon->IsFighterBay()))
+		return;
+	if(HasBays())
+	{
+		// Skip adding bay because it already exists.
+		for(unsigned int i = 0; i < bays.size(); i++)
+		{
+			// If bay exists for this index, then skip adding anything.
+			if(bays[i].bayIndex == bayIndex)
```

Comment:
`bayIndex` is necessary because some ships use the same X/Y coordinates for multiple guns.  One example is the Bastion which has multiple instances of 2 guns mounted on the same coordinates.  I added `bayIndex` for this very reason as a fix before publishing code.